### DISCUSSION TITLE
Add Store Hydration/Serialization API

### DISCUSSION THREAD

**Discussion by MikeRyanDev** [ngrx/platform#3078] [Ideas]
To better support state serialization and hydration I recommend adding built-in APIs into @ngrx/store. We should try and tackle the two biggest use cases:

* Serializing state for server side rendering
* Persisting state in browser and native platform storage

Ideally, we would separate the process of persistence and serialization into two steps:

1. *Transform* - Take a state object and perform a transformation on it
2. *Store* - Take the transformed result and save it

This would allow authors of persistence strategies to focus on persistence without having to reimplement key filtering or Immutable restructuring.

**Comment by MattiJarvinen** [comment]
Remember handling async storage in saving to indexedDB or some native cordova plugin filesystem etc.

**Comment by Matmo10** [comment]
Just throwing out ideas while we're on the topic ... but it would be awesome if something like https://github.com/btroncone/ngrx-store-localstorage became a first class part of ngrx itself. Serializing out for SSR and indexedDB usage is definitely important too, but I would guess that the majority of people are just saving bits of state to localstorage. Plus, the library above [doesn't work with feature modules](https://github.com/btroncone/ngrx-store-localstorage/issues/65).

**Comment by Dutchboy** [comment]
.... When is this going to be implemented? :-( Wasn't it for 7.0?? (today....)

**Comment by Dutchboy** [comment]
@brandonroberts When do you think it is going to work?

**Comment by itayod** [comment]
@brandonroberts may I take this?

**Comment by brandonroberts** [comment]
@Dutchboy @itayod we're going to put together a design for this first, which we have on our list to look at since version 8 has launched.

**Comment by itayod** [comment]
@brandonroberts sounds great!
meanwhile, I have posted a blog that might help users to implement it:
https://medium.com/better-programming/sync-your-state-in-local-storage-with-ngrx-9d6ceba93fc0

**Comment by biesbjerg** [comment]
I'm using [ngrx-store-sync](https://github.com/biesbjerg/ngrx-store-sync) that I hacked together. Seems to work for my use case. 

Putting it here to maybe give you some ideas how to do it (or not do it :D)

**Comment by rickvandermey** [comment]
@itayod maybe update your article and repo to 8, because of breaking changes in meta reducer

**Comment by itayod** [comment]
@rickvandermey Thanks for noticing me! but I think most of the users are still on version 7.X. so I guess update it now could lead to some confusion at this point :)

**Comment by LAlves91** [comment]
Any news on this one?

**Comment by classifieds-dev** [comment]
Is there any more news? What have you done to contribute.

**Comment by LAlves91** [comment]
@classifieds-dev relax man! I was just asking!
